### ISSUE TITLE
Render connection problems on last msg send causes render task to freeze

### ISSUE THREAD

**Issue by eberrippe** [CGRU/cgru#585]
Hi Timur, you remember our chat in the CGRU Telegram yesterday about our connection problems?

The problem is that the render does not notify the server that it finished the task. So the task stays at no progress but the server can also not eject it with  the "task no change timeout". We suppose its because the socket on the render freezes somehow cause of an connection issue.  Our only way of fixing is to exit the render client or restarting the afserver.

We were checking the code and we suppose that we found the issue we are running into.

What happens if the last output from the task was not sent successfully? We think there should be some kind of sleep and reconnection try going on here?

https://github.com/CGRU/cgru/blob/11e78f28060de14516d577c7a22f5ab6426f85cd/afanasy/src/render/renderhost.cpp#L274-L287

We imagine a solution could be sth like:
```
    while not ok {
        bool ok;
        af::Msg * server_answer = af::sendToServer( msg, ok,
            msg->type() == af::Msg::TRenderRegister ? af::VerboseOff : af::VerboseOn);
        if (ok)
            connectionEstablished();
        else
            serverUpdateFailed();
        delete msg;
        sleep(1)
        if retries > 5 {
            break;
        }
        m_up.clear();
        return server_answer;
        retry++;
        }
```
Or would this cause a total render lag?

Thanks a lot!

**Comment by timurhai** [comment]
Hello, Jan!
This function already runs in a cycle:
https://github.com/CGRU/cgru/blob/11e78f28060de14516d577c7a22f5ab6426f85cd/afanasy/src/render/main.cpp#L166
Cycle in a cycle not needed.

**Comment by ultra-sonic** [comment]
Hi Timur,

here are the most common log entries that we found on frozen afrender clients:
```
Mon 25 Sep 09:34.06: INFO    Failed to connect to server. Last success connect time: Mon 25 Sep 09:33.23
Mon 25 Sep 09:34.06: INFO    Last connect was: 43 seconds ago, Zombie time: 180s
AFERROR: msgsendtoaddress: connect failure for msgType 'TRenderUpdate':
10.10.8.150:51000: Connection refused
```
and
```
D=19970: Exit Code=0 Status=0
AFERROR: readdata: read: Resource temporarily unavailable

AFERROR: af::msgread: can't read message header, bytes = -1 (< Msg::SizeHeader).
AFERROR: msgsendtoaddress: Reading binary answer failed.
Tue 26 Sep 10:15.31: INFO    Failed to connect to server. Last success connect time: Tue 26 Sep 10:15.18
Tue 26 Sep 10:15.31: INFO    Last connect was: 13 seconds ago, Zombie time: 180s
Tue 26 Sep 10:45.21: ERROR   RenderHost::stopTask: No such task: TP: j171 b2 t422 (now running 0 tasks).
Tue 26 Sep 10:45.21: ERROR   RenderHost::stopTask: No such task: TP: j963 b3 t17 (now running 0 tasks).
```

Is there any way to get more verbose logging especially on what is going on with our sockets?

To me it would be interesting to get a log output every time RCVTIMEO and SNDTIMEO have been reached (on both ends -afserver and afrender) - I suspect this to be the cause of the `Resource temporarily unavailable` error

to debug `AFERROR: msgsendtoaddress: connect failure for msgType 'TRenderUpdate': 10.10.8.150:51000: Connection refused`
it would be great to know if the socket was in a `TIME_WAIT` state and therefore refused our connection attempt because the previous connection was still not properly terminated. is there a way to log more verbose info on `connect failure` as well?

I have a feeling that in our network we should increase RCVTIMEO and SNDTIMEO but not sure by how much? will this setting interfere with other timeouts like zombietime etc?

also please remind me where we can configure the delay that afrender waits between its connection attempts to the server. I think this should be longer then 120 secs because it may be that:
> connections will hold the TCP port in the TIME_WAIT state for 30-120 seconds
from https://stackoverflow.com/questions/3229860/what-is-the-meaning-of-so-reuseaddr-setsockopt-option-linux

Cheers
Oli

@sebastianelsner

**Comment by timurhai** [comment]
Hi, Oliver!
Afanasy uses macros to output system call errors:
https://github.com/CGRU/cgru/blob/master/afanasy/src/libafanasy/name_afnet.cpp#L247
Macros are defined there:
https://github.com/CGRU/cgru/blob/master/afanasy/src/include/macrooutput.h
And just uses perror to print a system error message:
https://man7.org/linux/man-pages/man3/perror.3.html
For now i do not way to get more detailed error explanation.
But may be someone, who more familiar with a Linux system calls can help.

**Comment by timurhai** [comment]
You can monitor sockets states by such command as `netstat`, `ss` and so on.
TIME-WAIT can be a problem:
https://cgru.readthedocs.io/en/latest/afanasy/server.html#time-wait
If so, may be you can reduce _Maximum Segment Lifetime_ ?

**Comment by timurhai** [comment]
Hello, Oli!
You want to say, that new task generate message that server can read and answer. And old task still can't connect to server? This can't happen, as render connects server once per cycle and send & recv one data. Different tasks fills data and then reads own portion. So, if render can connect to server each task can connect to server. Or you want to say that it is not?

**Comment by lithorus** [comment]
I just wants to add that we're seeing the exact same thing. It tends to happen more often when the server is under heavy load and tasks are fast to exit. My theory is that it starts and stops before the server acknowledges the starting of the task, but I have yet been able to re-produce the issue in a non-production system.

**Comment by lithorus** [comment]
@eberrippe Good idea about the cron job, although 30 min is a bit long for many of our tasks. But good to have as a fallback. Might implement the same here as a stop gap solution.

My theory is that it comes if there's packet loss and/or the server fails to process the "I'm done" request from the client, so the server still thinks it's doing things. One way to detect it aswell, is if you try to pull the output log of the task it will return something in the line of "Render has no task".

**Comment by timurhai** [comment]
Hello!
So, when render stucks, it starts to repeat this log:
```
AFERROR: af::msgread: can't read message header, bytes = -1 (< Msg::SizeHeader).
AFERROR: msgsendtoaddress: Reading binary answer failed.
Thu 07 Dec 14:45.11: INFO    Failed to connect to server. Last success connect time: Thu 07 Dec 14:44.57
Thu 07 Dec 14:45.11: INFO    Last connect was: 14 seconds ago, Zombie time: 180s
AFERROR: readdata: read: Resource temporarily unavailable
```
This means that this system call returns < 0:
https://github.com/CGRU/cgru/blob/1fd2bc4d198a848deaa509d5d502a15750b8fe29/afanasy/src/libafanasy/name_afnet.cpp#L42
And produces this system error message: `Resource temporarily unavailable`

What is server log this time?
What server "thinks" about this render and its tasks?
It does not became "zombie" and still runs latest task?
And task progress timeout does not works?

You want "stucked" render to exit manually to became zombie on server and later is can be started again?

**Comment by timurhai** [comment]
So, server log has errors, something unusual when renders stucks?

* Client side:
Have you tried to change socket oprions, to try to solve `Resource temporarily unavailable`?
You can increase `af_so_client_RCVTIMEO_sec` and `af_so_client_SNDTIMEO_sec` x2 - x3 times.
Also you can enable `af_so_client_TCP_NODELAY`, this way even very small data will be written.
You can read socket docs all over the web for socket options explanation.

Also you can try to use `recv` instead of `read` in this line:
https://github.com/CGRU/cgru/blob/1fd2bc4d198a848deaa509d5d502a15750b8fe29/afanasy/src/libafanasy/name_afnet.cpp#L42
May be the system error will be differ.

* Server side:
Do you use EPOLL or theading server? What is your `af_server_linux_epoll` option?
If you use theading, have you tried to increase `af_server_sockets_readwrite_threads_num`  x2 - x3 times?

**Comment by lithorus** [comment]
We have the same problem and ours are :
```
af_server_linux_epoll: 1
af_server_sockets_processing_threads_num: 30
af_server_sockets_readwrite_threads_num: 30
```
However we have the issue on another server with default options.

**Comment by timurhai** [comment]
Can you provide server logs on render stuck?

You can increase pool Heart Beat seconds for renders for a big farm (pool).
By default it is 1 second. So 1000 renders connects to server every second.
On a remote pool we set it to 5 seconds.

**Comment by timurhai** [comment]
Since pools appear (v >= 3.0.0) such render parameters as `af_render_zombietime`, `af_render_up_resources_period`, `af_render_heartbeat_sec` moved from config to pool parameters. This way we can configure them per pool.
So in your config such parameters has no effect.
Have you set them on pool? If no, you have 1 sec heartbeat on the entire farm.
Yes we can compare it with DDOS'ing.
But I still can't have not understand, why render can connect and write to server, but can't read. So it does not falls to zombie, but stuck, so wastes afserver (farm) time.

**Comment by ultra-sonic** [comment]
Hey Timur,

I don't know how familiar you are with the MQTT protocol and how it implements a so called Quality of Service (QoS).
This page describes it pretty well https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/

I am pretty sure that if we could implement at least "QoS 1" (server acknowledging reception of "task finished" to the client) it would solve this very issue.

What do you think?

**Comment by timurhai** [comment]
Now render updates server in a cycle.
It will send that task is finished or anything else till server says that it got it.

But as far as i understand the problem, render can't read any server answer (stuck situation).
So render can't receive any information from server, so server can't do anything with render too.

You say, that only afrender restart solves problem.
So afrender should "decide" to exit manually, am i right?
We can count such socket read fails, for example.

**Comment by ultra-sonic** [comment]
no restarts please...we must fix the root cause and not patch the issue

**Comment by lithorus** [comment]
I think the first step is to be able to re-produce it in a reliable way.

**Comment by ultra-sonic** [comment]
correct! the task is stuck and the client happily accepts new tasks and renders them as if nothing ever happened.
The task is not DON on the server - it is still RUN

**Comment by lithorus** [comment]
Most of them time they are at not even at 0%. That's how I usually detect the. They have started, but don't have a % yet.
